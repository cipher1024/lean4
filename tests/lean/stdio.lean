
#eval do
out ← IO.getStdout;
(out.putStrLn "print stdout" : IO Unit)

#eval do
err ← IO.getStderr;
(err.putStr "print stderr" : IO Unit)

open IO

def test : IO Unit := do
⟨ instr, outstr ⟩ ← Pipe.mk true;
FS.withFile "stdout1.txt" IO.FS.Mode.write $ fun h₁ => do
{ h₂ ← FS.Handle.mk "stdout2.txt" IO.FS.Mode.write;
  println "foo";
  withStdout h₁ $ do
    println "line 1";
    withStdout outstr $ do
    { println "line 7";
      catch
      ( do
        withStdout h₂ $ println "line 2";
        throw $ IO.userError "my error" )
      ( fun e => println e ) };
    println "line 3" };
println "line 4";
println "\n> stdout1.txt";
readFile "stdout1.txt" >>= print;
println "\n> stdout2.txt";
readFile "stdout2.txt" >>= print;
println "\n> >= pipe =>";
FS.Handle.readToExhaustion instr >>= print;
outstr.putStrLn "more stuff";
outstr.putStrLn "and some more";
FS.Handle.readToExhaustion instr >>= print;
pure ()

#eval test
