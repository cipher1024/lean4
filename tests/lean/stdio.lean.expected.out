print stdout
print stderr
foo
line 4

> stdout1.txt
line 1
line 3

> stdout2.txt
line 2

> >= pipe =>
line 7
my error
more stuff
and some more
